src/bunit.web.query/Labels/LabelQueryExtensions.cs

```diff
@@ -0,0 +1,52 @@
+using System.Collections.ObjectModel;
+using AngleSharp.Dom;
+using Bunit.Labels.Strategies;
+
+namespace Bunit;
+
+public static class LabelQueryExtensions
+{
+	private static readonly IReadOnlyList<ILabelTextQueryStrategy> LabelTextQueryStrategies = new ILabelTextQueryStrategy[]
+	{
+		// This is intentionally in the order of most likely to minimize strategies tried to find the label
+		new LabelTextUsingForAttributeStrategy(),
+		new LabelTextUsingAriaLabelStrategy(),
+		new LabelTextUsingWrappedElementStrategy(),
+		new LabelTextUsingAriaLabelledByStrategy(),
+	};
+
+	/// <summary>
+	/// Returns the first element (i.e. an input, select, textarea, etc. element) associated with the given label text.
+	/// </summary>
+	/// <param name="renderedFragment">The rendered fragment to search.</param>
+	/// <param name="labelText">The text of the label to search (i.e. the InnerText of the Label, such as "First Name" for a `<label>First Name</label>`)</param>
+	public static IElement FindByLabelText(this IRenderedFragment renderedFragment, string labelText)
+	{
+		return FindByLabelTextInternal(renderedFragment, labelText) ?? throw new LabelNotFoundException(labelText);
+	}
+
+	/// <summary>
+	/// Returns the first element (i.e. an input, select, textarea, etc. element) associated with the given label text.
+	/// </summary>
+	/// <param name="renderedFragment">The rendered fragment to search.</param>
+	/// <param name="labelText">The text of the label to search (i.e. the InnerText of the Label, such as "First Name" for a `<label>First Name</label>`)</param>
+	internal static IElement? FindByLabelTextInternal(this IRenderedFragment renderedFragment, string labelText)
+	{
+		try
+		{
+			foreach (var strategy in LabelTextQueryStrategies)
+			{
+				var element = strategy.FindElement(renderedFragment, labelText);
+
+				if (element != null)
+					return element;
+			}
+		}
+		catch (DomException exception) when (exception.Message == "The string did not match the expected pattern.")
```

Comment:
We can make this a bit more stable, in theory, but comparing the error code instead:

```suggestion
		catch (DomException exception) when (exception.Code == (int)DomError.Syntax)
```

Reply:
Another option is to use lower-level AngleSharp APIs instead and not have the exception-handling logic at all, which I am more in favor of. E.g., we can add the following extensions methods:

```c#
internal static IElement? TryQuerySelector(this INodeList nodes, string cssSelector)
{
	if (nodes.Length == 0 || nodes[0].Owner?.Context.GetService<ICssSelectorParser>() is not ICssSelectorParser cssParser)
		return null;

	if (cssParser.ParseSelector(cssSelector) is not ISelector selector)
		return null;

	return nodes.QuerySelector(selector);
}

internal static IEnumerable<IElement> TryQuerySelectorAll(this INodeList nodes, string cssSelector)
{
	if (nodes.Length == 0 || nodes[0].Owner?.Context.GetService<ICssSelectorParser>() is not ICssSelectorParser cssParser)
		return Enumerable.Empty<IElement>();

	if (cssParser.ParseSelector(cssSelector) is not ISelector selector)
		return Enumerable.Empty<IElement>();

	return nodes.QuerySelectorAll(selector);
}
```

And then everywhere where we are using `QuerySelector` or `QuerySelectorAll`, call `TryQuerySelector` or `TryQuerySelectorAll` instead.

Reply:
Done in 0123653